### ISSUE TITLE
`cargo clippy --fix` fails on a codebase that compiles (`cargo build` succeeds)

### ISSUE THREAD

**Issue by 0x53A** [rust-lang/rust-clippy#14319]
### Summary

```
> cargo build
[...]
warning: `crab-fs` (bin "crab-fs") generated 123 warnings (run `cargo fix --bin "crab-fs"` to apply 61 suggestions)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1m 15s
```

```
> cargo clippy --fix --allow-dirty
[...]
    Checking crab-fs v0.1.0 (/mnt/d/repos/crab-fs)
warning: failed to automatically apply fixes suggested by rustc to crate `crab_fs`

after fixes were automatically applied the compiler reported errors within these files:

  * src/cuttlefish.rs
  * src/errors.rs
  * src/io/encrypted_fs.rs
  * src/io/fs.rs
  * src/main.rs
  * src/repository/repository_v1.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
[...]
error[E0596]: cannot borrow `dst_file` as mutable, as it is not declared as mutable
   --> src/io/encrypted_fs.rs:660:9
    |
660 |         dst_file.seek(SeekFrom::Start(dst_pos))?;
    |         ^^^^^^^^ cannot borrow as mutable
    |
[...]
error: aborting due to 3 previous errors; 55 warnings emitted
```

---------------------------

Interestingly, it works with ``--broken-code``, and I've committed the result at https://github.com/0x53A/crab-fs/commit/b1c3e64442f46d156f9703b9889fcaaa62399088.

That is, the following sequence fails ...

```
git checkout 3721927f7333c3966b58269e2f656779211dd30d
cargo clippy --fix --allow-dirty

[...]
error: aborting due to 3 previous errors; 55 warnings emitted
[...]
```

... while this works:

```
git checkout 3721927f7333c3966b58269e2f656779211dd30d
cargo clippy --fix --allow-dirty --broken-code
cargo clippy --fix --allow-dirty
```

I've committed all files while ``cargo clippy --fix`` (without the ``--broken-code``) was running, to be able to see what's the difference and have committed it as https://github.com/0x53A/crab-fs/commit/0539970969d8eb17398af1cd6b22e39e87e711b7.

Running ``cargo build`` on 0539970969d8eb17398af1cd6b22e39e87e711b7 also fails with the same three errors.

There is the delta between these two:

https://github.com/0x53A/crab-fs/compare/b1c3e64442f46d156f9703b9889fcaaa62399088..0539970969d8eb17398af1cd6b22e39e87e711b7

--------------------

The full log is attached

[full_log.txt](https://github.com/user-attachments/files/19019971/full_log.txt)

### Reproducer

The repository is here: https://github.com/0x53A/crab-fs

```
>git rev-parse head
3721927f7333c3966b58269e2f656779211dd30d
```

To reproduce the error, you need to check out that specific commit

```
git checkout 3721927f7333c3966b58269e2f656779211dd30d
cargo clippy --fix --allow-dirty

[...]
error: aborting due to 3 previous errors; 55 warnings emitted
[...]
```

### Version

```text
> rustc -Vv

rustc 1.87.0-nightly (96cfc7558 2025-02-27)
binary: rustc
commit-hash: 96cfc75584359ae7ad11cc45968059f29e7b44b7
commit-date: 2025-02-27
host: x86_64-unknown-linux-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0

-------------------

> rustup -vV
rustup 1.27.1 (54dd3d00f 2024-04-24)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.87.0-nightly (96cfc7558 2025-02-27)`

------------------

> cargo -vV
cargo 1.87.0-nightly (1d1d646c0 2025-02-21)
release: 1.87.0-nightly
commit-hash: 1d1d646c06a84c1aa53967b394b7f1218f85db82
commit-date: 2025-02-21
host: x86_64-unknown-linux-gnu
libgit2: 1.9.0 (sys:0.20.0 vendored)
libcurl: 8.12.0-DEV (sys:0.4.79+curl-8.12.0 vendored ssl:OpenSSL/3.4.0)
ssl: OpenSSL 3.4.0 22 Oct 2024
os: Ubuntu 22.4.0 (jammy) [64-bit]
```

**Comment by KonaeAkira** [comment]
Clippy suggests changing this line:
```rust
(&*file).seek(SeekFrom::Start(current_pos))?;
```
to this line as part of the `borrow_deref_ref` lint:
```rust
file.seek(SeekFrom::Start(current_pos))?; // does not compile
```
Possible duplicate of https://github.com/rust-lang/rust-clippy/issues/9905?
